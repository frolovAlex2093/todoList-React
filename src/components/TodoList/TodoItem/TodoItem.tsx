import React from "react";
import { IconButton, Box, Paper, Typography } from "@mui/material";
import { Delete as DeleteIcon, Edit as EditIcon, MoreVert as MoreVertIcon } from "@mui/icons-material";

import type { Todo } from "../../../pages/TodoHome";
import { Link } from "react-router-dom";

interface TodoItemProps {
  todo: Todo;
  onDeleteTodo: (id: Todo["id"]) => void;
  onCheckTodo: (id: Todo["id"]) => void;
  onEdit: (id: Todo["id"]) => void;
  onClickMore: (todo: Todo) => void
}

export const TodoItem: React.FC<TodoItemProps> = ({
  todo,
  onDeleteTodo,
  onCheckTodo,
  onEdit,
  onClickMore
}) => (
  <Paper
    elevation={1}
    sx={{
      marginBottom: "15px",
      width: "100%",
      padding: "15px 20px",
      borderRadius: 1,
      gap: 2,
      opacity: todo.checked ? 0.5 : 1,
    }}
  >
    <Box textAlign="left">
      <Typography
        onClick={() => onCheckTodo(todo.id)}
        sx={{
          cursor: "pointer",
          textDecorationLine: todo.checked ? "line-through" : "none",
        }}
        variant="h5"
        component="h5"
        gutterBottom
      >
        {todo.name}
      </Typography>
      <Typography
        variant="subtitle1"
        component="div"
        gutterBottom
        sx={{
          whiteSpace: "nowrap",
          overflow: "hidden",
          textOverflow: "ellipsis",
        }}
      >
        {todo.description}
      </Typography>
    </Box>
    <Box display="flex" justifyContent="flex-end">
      <Link to="/TodoAbout">
        <IconButton color="primary" aria-label="edit" onClick={() => onClickMore(todo)}>
          <MoreVertIcon />
        </IconButton>
      </Link>
      <IconButton
        onClick={() => onEdit(todo.id)}
        color="primary"
        aria-label="edit"
      >
        <EditIcon />
      </IconButton>
      <IconButton
        onClick={() => onDeleteTodo(todo.id)}
        color="error"
        aria-label="delete"
      >
        <DeleteIcon />
      </IconButton>
    </Box>
  </Paper>
);
