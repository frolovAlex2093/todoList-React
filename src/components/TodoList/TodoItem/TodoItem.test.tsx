import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import { TodoItem } from "./TodoItem";

const todoTest = {
  id: 1,
  name: "test name",
  description: "test description",
  checked: false,
};

const onDeleteTodoTest = jest.fn();
const onCheckTodoTest = jest.fn();
const onEditTest = jest.fn();
const onClickMoreTest = jest.fn();

describe("TodoItem component", () => {
  it("TodoItem render", () => {
    render(
      <MemoryRouter>
        <TodoItem
          todo={todoTest}
          onDeleteTodo={onDeleteTodoTest}
          onCheckTodo={onCheckTodoTest}
          onEdit={onEditTest}
          onClickMore={onClickMoreTest}
        />
      </MemoryRouter>
    );

    expect(screen.getByText("test name")).toBeInTheDocument();
    expect(screen.getByText("test description")).toBeInTheDocument();
  });

  it("buttons work", () => {
    render(
      <MemoryRouter>
        <TodoItem
          todo={todoTest}
          onDeleteTodo={onDeleteTodoTest}
          onCheckTodo={onCheckTodoTest}
          onEdit={onEditTest}
          onClickMore={onClickMoreTest}
        />
      </MemoryRouter>
    );

    userEvent.click(screen.getByRole("button", { name: "delete" }));
    userEvent.click(screen.getByRole("button", { name: "edit" }));
    userEvent.click(screen.getByRole("button", { name: "more" }));
    userEvent.click(screen.getByText("test name"));

    expect(onDeleteTodoTest).toHaveBeenCalled();
    expect(onCheckTodoTest).toHaveBeenCalled();
    expect(onEditTest).toHaveBeenCalled();
    expect(onClickMoreTest).toHaveBeenCalled();
  });

  it("TodoItem snapshot", () => {
    const todoItem = render(
        <MemoryRouter>
          <TodoItem
            todo={todoTest}
            onDeleteTodo={onDeleteTodoTest}
            onCheckTodo={onCheckTodoTest}
            onEdit={onEditTest}
            onClickMore={onClickMoreTest}
          />
        </MemoryRouter>
      );

      expect(todoItem).toMatchSnapshot()
  })
});
