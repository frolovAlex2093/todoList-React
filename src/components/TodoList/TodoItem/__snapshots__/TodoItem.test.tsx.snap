// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem component TodoItem snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1sl6q6t-MuiPaper-root"
      >
        <div
          class="MuiBox-root css-13brihr"
        >
          <h5
            class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-1bju5vi-MuiTypography-root"
          >
            test name
          </h5>
          <div
            class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-gutterBottom css-goc1fa-MuiTypography-root"
          >
            test description
          </div>
        </div>
        <div
          class="MuiBox-root css-1bvc4cc"
        >
          <a
            href="/TodoAbout"
          >
            <button
              aria-label="more"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeMedium css-1kuq5xv-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="MoreVertIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </a>
          <button
            aria-label="edit"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeMedium css-1kuq5xv-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="EditIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            aria-label="delete"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorError MuiIconButton-sizeMedium css-1jo1f9u-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="DeleteIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1sl6q6t-MuiPaper-root"
    >
      <div
        class="MuiBox-root css-13brihr"
      >
        <h5
          class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-1bju5vi-MuiTypography-root"
        >
          test name
        </h5>
        <div
          class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-gutterBottom css-goc1fa-MuiTypography-root"
        >
          test description
        </div>
      </div>
      <div
        class="MuiBox-root css-1bvc4cc"
      >
        <a
          href="/TodoAbout"
        >
          <button
            aria-label="more"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeMedium css-1kuq5xv-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="MoreVertIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </a>
        <button
          aria-label="edit"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeMedium css-1kuq5xv-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="EditIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          aria-label="delete"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorError MuiIconButton-sizeMedium css-1jo1f9u-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="DeleteIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
